### DISCUSSION TITLE
Deadlocks in Qrtz_Triggers table

### DISCUSSION THREAD

**Discussion by nareshj1007** [quartznet/quartznet#1254] [Q&A] [answered]
If this is a question, please use the [discusssion feature](https://github.com/quartznet/quartznet/discussions)
We use Quartz.net 2.3.3 version for job scheduling and continuously seeing deadlocks in QRTZ_TRIGGERS table in Quartz database.. Below is the deadlock captured on quartz database. Please see attached deadlocks graph for more details.

[Quartz_Deadlocks.txt](https://github.com/quartznet/quartznet/files/6876477/Quartz_Deadlocks.txt)

We use Microsoft SQL Server database for Quartz DB and have a clustered environment with two nodes to execute the jobs. Here is the quartz.config file entries.

quartz.scheduler.instanceName = ESBScheduler
quartz.scheduler.instanceId = AUTO
quartz.threadPool.type = Quartz.Simpl.SimpleThreadPool, Quartz
quartz.threadPool.threadCount = 40
quartz.threadPool.threadPriority = Normal
quartz.jobStore.type = Quartz.Impl.AdoJobStore.JobStoreTX, Quartz
quartz.jobStore.dataSource = default
quartz.jobStore.tablePrefix = QRTZ_
quartz.jobStore.clustered = true
org.quartz.jobStore.clusterCheckinInterval = 20000
quartz.jobStore.lockHandler.type = Quartz.Impl.AdoJobStore.SimpleSemaphore, Quartz
quartz.jobStore.driverDelegateType = Quartz.Impl.AdoJobStore.StdAdoDelegate, Quartz
quartz.dataSource.default.provider = SqlServer-20
quartz.jobStore.useProperties = false
quartz.jobStore.selectWithLockSQL = SELECT * FROM {0}LOCKS WITH (UPDLOCK, ROWLOCK) WHERE SCHED_NAME = {1} AND LOCK_NAME = @lockname
quartz.dataSource.default.connectionStringName = XXXXXX

Please let me know if I am missing any configuration and request your help in resolving these deadlocks.

**Comment by lahma** [comment, marked as answer]
* You are using a very old version of Quartz, you should be able to upgrade to 2.6.2 without major code changes, just follow the release notes - there's one DB column addition needed
* You are using `StdAdoDelegate`, you should use `SqlServerDelegate`
* Can' remember whether `selectWithLockSql` came automatic with 3.x series or was it already in 2.x, Quartz can nowadays select the optimal for you based on delegate/provider
* You should probably close [the issue](https://github.com/quartz-scheduler/quartz/issues/700) you opened on Java side, cross-posting is uncool

Ideally you would upgrade all the way to 3.x to get better support story - but there are API changes etc. But [all documented](https://www.quartz-scheduler.net/documentation/quartz-3.x/migration-guide.html).

**Reply by nareshj1007** [reply]
Thanks a lot Iahma.

We made suggested changes and monitored it for two days and didn't see any deadlocks on Qrtz_Triggers table so far. So, it looks the recommended change is working. 

We will monitor for few more days for any deadlocks and will let you know. Thanks again.

**Reply by nareshj1007** [reply]
We monitored it for few more days and there is no sign of deadlock on Qrtz_Triggers table now.  Thanks.

**Reply by lahma** [reply]
Great to hear,  thanks for the update!

**Comment by nareshj1007** [comment]
Thanks Iahma for the inputs.

We surely can consider upgrading to latest version as a long term solution. Is there any immediate solution/workaround in 2.3.3 version to prevent these deadlocks?

**Reply by lahma** [reply]
Maybe check that indexes are valid, 3.x has performance improvements and all kinds of work done. Maybe check 2.6.2 migration first and release notes if any fix matches what you've been experiencing. Custom lock handler SQL might be related...
